Basic Verilator infrastructure (#943)
cocotb/cocotb: documentation/source/simulator_support.rst

See ``access_single_bit`` test in :file:`examples/functionality/tests/test_discovery.py`.

Verilator
---------

cocotb supports Verilator 4.020 and above.

> Verilator converts Verilog code to C++ code that is compiled into highly optimized simulation and does not support VHDL.

Reviewer: "compiled into highly optimized simulation" sounds a bit awkward, but I'm not sure what would be better.

Reviewer 2: "that is compiled, resulting in highly optimized simulation"?

> One major limitation to standard Verilog simulators is that it does not support delays assignments.

Reviewer: ```suggestion
One major limitation compared to standard Verilog simulators is that it does not support delay assignments.
```

> To run cocotb with Verilator, you need verilator in your path or it can be discoverd by pkg-config.

Reviewer: ```suggestion
To run cocotb with Verilator, you need `verilator` in your path or it can be discovered by `pkg-config`.
```

Finally, cocotb currently generates a Verilator toplevel C++ simulation loop which is timed at the highest precision.

> If your design's clocks vary in precision, the performance of the simulation can be improved in the same order of magnitude by adjusting the precision in the Makefile, e.g.,

Reviewer: ```suggestion
If your designs clocks vary in precision, the performance of the simulation can be improved in the same order of magnitude by adjusting the precision in the Makefile, e.g.,
```

See ``access_single_bit`` test in :file:`examples/functionality/tests/test_discovery.py`.

Verilator
---------

cocotb supports Verilator 4.020 and above.
Verilator converts Verilog code to C++ code that is compiled.
It does not support VHDL.

> One major limitation compared to standard Verilog simulators is that it does not support delays assignments.

Reviewer: ```suggestion
One major limitation compared to standard Verilog simulators is that it does not support delay assignments.
```

Reviewer 3: Or perhaps `delayed`?

> To run cocotb with Verilator, you need `verilator` in your path or it can be discoverd by pkg-config.

Reviewer: ```suggestion
To run cocotb with Verilator, you need `verilator` in your path or it can be discovered by `pkg-config`.
```

Reviewer 3: Needs double-backticks throughout

Finally, cocotb currently generates a Verilator toplevel C++ simulation loop which is timed at the highest precision.
If your design's clocks vary in precision, the performance of the simulation can be improved in the same order of magnitude by adjusting the precision in the Makefile, e.g.,

.. code-block:: makefile

>     PRECISION=6 # Set precision to 10^-6s

Reviewer: Reflect the naming change here. Also, wouldn't it have been "-6" (https://github.com/chipsalliance/verilator/blob/master/include/verilated.h#L709)?

> 

Reviewer: ```suggestion
.. versionadded:: 1.3
```
